### PR TITLE
Selfie: Implement aspect ratio selection

### LINKED ISSUE #18: Implement view photo ratio
We should be able to switch between wide crop ( 16:9 ) and full image ( 4:3 )

### PR DESCRIPTION
**Pull request by SebaUbuntu**
Fixes #18

Change-Id: I07c4ba1c06ce5009862b422904d6f4587e742332

### CODE REVIEW

**Comment by SebaUbuntu** [comment]
TODO
- We need to change the icon but I can't find good drawables for 1:1, 4:3 and 16:9
- ~PreviewView is currently above top bar~

File: app/src/main/res/layout/activity_main.xml [hunk 1]
```diff
@@ -9,6 +9,77 @@
...
+        android:layout_height="0dp"
+        app:layout_constraintBottom_toTopOf="@+id/bottomButtonsLayout"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/topButtonsLayout">
+
+    </androidx.camera.view.PreviewView>
+
+
+    <ImageView
```
Comment (luca020400) [inline app/src/main/res/layout/activity_main.xml]: That's unrelated, but shouldn't we put the items known to be show over the preview inside the view?

File: app/src/main/java/org/lineageos/selfie/MainActivity.kt [hunk 2]
```diff
@@ -392,6 +399,14 @@ class MainActivity : AppCompatActivity() {
         cameraController.cameraSelector = cameraSelector
         cameraController.setEnabledUseCases(cameraUseCases)
 
+        // Set aspect ratio
+        aspectRatio = sharedPreferences.aspectRatio
+        if (oldY == -1f) {
+            oldY = viewFinder.y
+        }
+        viewFinder.layoutParams.height =
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/MainActivity.kt]: Wait, this just change the view finder height
What about the actual output?

Comment (SebaUbuntu) [reply app/src/main/java/org/lineageos/selfie/MainActivity.kt]: CameraController will adapt the resolution/aspect ratio based on the preview view aspect ratio
In fact with 4:3 I get a 3000x4000 image, with 16:9 a 2250x4000 image and with 1:1 a 3000x3000 image

File: app/src/main/res/layout/activity_main.xml (hunk 1 above)
Comment (SebaUbuntu) [reply app/src/main/res/layout/activity_main.xml]: Just tried, for somehow reason findViewById doesn't work for the elements inside the view (returns null)

**Comment by luk1337** [comment]
Maybe ToggleButton would be good enough instead of drawable:
```diff
diff --git a/app/src/main/java/org/lineageos/selfie/MainActivity.kt b/app/src/main/java/org/lineageos/selfie/MainActivity.kt
index 9cd1b61..b2af2ec 100644
--- a/app/src/main/java/org/lineageos/selfie/MainActivity.kt
+++ b/app/src/main/java/org/lineageos/selfie/MainActivity.kt
@@ -29,6 +29,7 @@ import android.view.View
 import android.view.ViewConfiguration
 import android.widget.ImageView
 import android.widget.Toast
+import android.widget.ToggleButton
 import androidx.appcompat.app.AppCompatActivity
 import androidx.camera.core.CameraSelector
 import androidx.camera.core.ImageCapture
@@ -70,7 +71,7 @@ import java.util.Timer
 import java.util.TimerTask
 
 class MainActivity : AppCompatActivity() {
-    private val aspectRatioButton by lazy { findViewById<ImageView>(R.id.aspectRatioButton) }
+    private val aspectRatioButton by lazy { findViewById<ToggleButton>(R.id.aspectRatioButton) }
     private val effectButton by lazy { findViewById<ImageView>(R.id.effectButton) }
     private val flashButton by lazy { findViewById<ImageView>(R.id.flashButton) }
     private val flipCameraButton by lazy { findViewById<ImageView>(R.id.flipCameraButton) }
@@ -489,6 +490,7 @@ class MainActivity : AppCompatActivity() {
         updateCameraModeButtons()
         toggleRecordingChipVisibility()
         updateTimerModeIcon()
+        updateAspectRatioIcon()
         updatePhotoEffectIcon()
         updateGridIcon()
         updateFlashModeIcon()
@@ -556,6 +558,10 @@ class MainActivity : AppCompatActivity() {
         bindCameraUseCases()
     }
 
+    private fun updateAspectRatioIcon() {
+        aspectRatioButton.text = sharedPreferences.aspectRatio.toString()
+    }
+
     /**
      * Update the grid button icon based on the value set in grid view
      */
diff --git a/app/src/main/java/org/lineageos/selfie/utils/AspectRatio.kt b/app/src/main/java/org/lineageos/selfie/utils/AspectRatio.kt
index c3f13ed..f72363a 100644
--- a/app/src/main/java/org/lineageos/selfie/utils/AspectRatio.kt
+++ b/app/src/main/java/org/lineageos/selfie/utils/AspectRatio.kt
@@ -10,6 +10,10 @@ package org.lineageos.selfie.utils
  * A dumb aspect ratio typing
  */
 class AspectRatio(private val h: Int, private val w: Int) {
+    override fun toString(): String {
+        return "$h:$w"
+    }
+
     fun getRatio(): Double {
         return h.toDouble() / w
     }
diff --git a/app/src/main/res/drawable/ic_crop.xml b/app/src/main/res/drawable/ic_crop.xml
deleted file mode 100644
index 8e91f26..0000000
--- a/app/src/main/res/drawable/ic_crop.xml
+++ /dev/null
@@ -1,5 +0,0 @@
-<vector android:height="24dp" android:tint="#000000"
-    android:viewportHeight="24" android:viewportWidth="24"
-    android:width="24dp" xmlns:android="http://schemas.android.com/apk/res/android">
-    <path android:fillColor="@android:color/white" android:pathData="M17,15h2V7c0,-1.1 -0.9,-2 -2,-2H9v2h8v8zM7,17V1H5v4H1v2h4v10c0,1.1 0.9,2 2,2h10v4h2v-4h4v-2H7z"/>
-</vector>
diff --git a/app/src/main/res/layout/activity_main.xml b/app/src/main/res/layout/activity_main.xml
index b2f08bb..5b1bd1f 100644
--- a/app/src/main/res/layout/activity_main.xml
+++ b/app/src/main/res/layout/activity_main.xml
@@ -107,14 +107,13 @@
             app:layout_constraintTop_toTopOf="parent"
             app:layout_constraintVertical_bias="0.0" />
 
-        <ImageButton
+        <ToggleButton
             android:id="@+id/aspectRatioButton"
             style="@style/SelfieTopBarButton"
             android:layout_width="@dimen/top_view_buttons_size"
             android:layout_height="@dimen/top_view_buttons_size"
             android:background="?attr/selectableItemBackgroundBorderless"
             android:contentDescription="@string/aspect_ratio_button_description"
-            android:src="@drawable/ic_crop"
             app:layout_constraintBottom_toBottomOf="parent"
             app:layout_constraintEnd_toStartOf="@+id/effectButton"
             app:layout_constraintHorizontal_bias="0.5"
```

**Comment by luca020400** [comment]
```
>>> 2268/4032
0.5625
>>> 9/16
0.5625
>>> 3060/4080
0.75
>>> 3/4
0.75
>>>
```

This is what I see on here
Google camera somehow does 3072/4080 but I'm not going to look into that...

**Review by luca020400 — APPROVED** [review]

**Review by luk1337 — CHANGES_REQUESTED** [review]
still no icon

**Review by luk1337 — APPROVED** [review]

**Review by luca020400 — APPROVED** [review]

**Pull request merged**